### ISSUE TITLE
arrow keys no longer scroll in less in Windows Terminal

### ISSUE THREAD

**Issue by jeremyd2019** [git-for-windows/git#3657]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.35.0.windows.1
cpu: i686
built from commit: 853373155dce195a91e6e41a58c8b8e5c5a94652
sizeof-long: 4
sizeof-size_t: 4
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19043.1466]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VIM
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: WinSSL
CRLF Option: CRLFAlways
Bash Terminal Option: ConHost
Git Pull Behavior Option: FFOnly
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Git Bash in Windows Terminal

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
less somefile
```
(actually noticed in `git diff`/`git show` when git runs the pager, but it reproduces without needing to be invoked through git)

 - What did you expect to occur after running these commands?

Able to scroll in `less` with arrow keys

 - What actually happened instead?

Arrow keys cause a bell, have to scroll with `j`/`k` for single line scrolling instead.  This apparently stopped working when I upgraded from 2.34.1 to 2.35.0.  Most likely an msys2-runtime issue, but it works fine with latest upstream msys2-runtime and msys2's less.

**Comment by jeremyd2019** [comment]
This also occurs in Git Bash in Windows conhost.

**Comment by bbbush** [comment]
sounds like a termcap issue? Other than arrow keys not working, ctrl-c also interrupts a SSH session, in an IntelliJ terminal.

**Comment by dscho** [comment]
I would agree that this most likely is due to the upgrade of the MSYS2 runtime to v3.3.3 (including [an ugly hack](https://github.com/git-for-windows/msys2-runtime/commit/1d0701d84e198b22f752bceda2fb73666029761f) to fix the now pretty thoroughly broken `disable_pcon` mode).

Can you verify whether creating `/etc/git-bash.config` containing the line `MSYS=enable_pcon` works around the problem?

Of course, then you incur all the nasty bugs in the `enable_pcon` mode. I really wish that Cygwin would have been more careful to leave the `disable_pcon` code path essentially untouched.

**Comment by albertosantini** [comment]
About `ctrl-c also interrupts a SSH session` the behaviour it is like ssh is invoked with `-t` option. From ssh man:

```
     -t      Force pseudo-terminal allocation.  This can be used to
             execute arbitrary screen-based programs on a remote
             machine, which can be very useful, e.g. when implementing
             menu services.  Multiple -t options force tty allocation,
             even if ssh has no local tty.
```

From https://unix.stackexchange.com/questions/102061/ctrl-c-handling-in-ssh-session#161878:
```
ssh -t remotehost command args ... will run an interactive session on remotehost, 
just like the above, except on the remote side, your_shell -c "command args ..." will be run. 
As above, if you type Ctrl-C, it will be sent to the remote host, where the command 
will likely receive SIGINT and immediately exit, and then the remote shell will exit. 
The remote sshd then closes the connection, and ssh reports Connection to remotehost closed.
```

As workaround for this, I have been using `alias ssh="/c/Windows/System32/OpenSSH/ssh.exe"`

Adding `export MSYS=enable_pcon` (I cannot found `/etc/git-bash.config` in the portable configuration), as suggested by @dscho, work arounds the problem. Thanks.

**Comment by dscho** [comment]
> I cannot found `/etc/git-bash.config` in the portable configuration

Right, you need to create this file.

**Comment by IllusionMH** [comment]
I had problems with keys behavior:
* <kbd>Home</kbd> did nothing 
* <kbd>End</kbd> printed `~`
* (testet after seeing this issue) <kbd>ArrowUp</kbd>/<kbd>ArrowDown</kbd> didn't scroll output of `less`/`git diff`

Creating `C:\Program Files\Git\etc\git-bash.config` with `MSYS=enable_pcon` fixed this problem for me.

**Comment by jeremyd2019** [comment]
Several others beat me to it, so probably just piling on at this point, but `enable_pcon` does indeed work around it for me too.

**Comment by ml-evb** [comment]
above workaround worked for me till 2.35.2 release. Now it looks like `MSYS=enable_pcon` is ignored.

**Comment by parktheredcar** [comment]
Came here trying to find out why ctrl-c is killing ssh connections, didn't see this problem in `git version 2.29.2.windows.2`. 

adding `export MSYS=enable_pcon` to `~/.bashrc` is currently working for me, tested with new windows terminal

`Windows Terminal Preview Version: 1.12.3472.0`
`git version 2.35.1.windows.2`

**Comment by jeremyd2019** [comment]
I'm not sure the ctrl-c killing ssh issue is really related to the apparent key mapping/termcap issue.  I don't see the key issues in upstream msys2-runtime, but the ssh issue has been reported there (msys2/msys2-runtime#83).

**Comment by ml-evb** [comment]
> adding export MSYS=enable_pcon to ~/.bashrc is currently working for me

thanks for the tip. Works here also. What stopped to work is putting `MSYS=enable_pcon` in `C:\Program Files\Git\etc\git-bash.config`

**Comment by CrendKing** [comment]
I had problem with `tig`. Worked around this by simply adding `MSYS=enable_pcon` to environment variables.

**Comment by ffes** [comment]
Indeed adding `export MSYS=enable_pcon` to my `~/.bashrc` helps to workaround this problem. I ran into it using `tig`.

But does anyone have any idea how I combine two MSYS settings?

I've tried `export MSYS="winsymlinks:nativestrict,enable_pcon"` and `export MSYS="winsymlinks:nativestrict;enable_pcon"` but that doesn't work. What separator should I use?

**Comment by rimrul** [comment]
I think the expected separator is a space.

**Comment by dscho** [comment]
> What stopped to work is putting `MSYS=enable_pcon` in `C:\Program Files\Git\etc\git-bash.config`

@ml-evb are you sure that it was not overwritten by the Git for Windows installer? You need to check the "enable experimental support for pseudo consoles" checkbox (it might be phrased slightly differently, but I trust you to recognize it when you see it).

**Comment by ml-evb** [comment]
@dscho that might be. Adding `export MSYS=enable_pcon` to `.bashrc` is a better option then.

**Comment by dscho** [comment]
> Is this workaround required for all future Git-For-Windows versions?

Hopefully not. There is a troublesome problem here: this workaround that you need right now is apparently triggered by a fix we _require_ to keep Git Credential Manager working in non-pcon mode. The underlying bug is really in the Cygwin runtime, but it is not a single bug. The entire support for pseudo consoles is unnecessarily convoluted and full of hacks and special-casing specific scenarios.

**Comment by jeremyd2019** [comment]
https://github.com/msys2/msys2-runtime/commit/8a95249d738cd46299c72a5da8150b4bba0a7e95 touches similar code, maybe some faint hope that this might be related?

**Comment by rimrul** [comment]
> https://github.com/msys2/msys2-runtime/commit/8a95249d738cd46299c72a5da8150b4bba0a7e95 touches similar code, maybe some faint hope that this might be related?

I don't think it is. Looks related to #3607, though.

**Comment by matteocoder** [comment]
@ffes I ran into the same problem, and what worked for me was creating a file `C:\Program Files\Git\etc\git-bash.config`
with the following content:
```
MSYS=enable_pcon winsymlinks:nativestrict
```
Notice that the two options are separated by a space and the variable is *not* quoted.
This way I can enable both pcon support and I am able to create real symlinks with `ln`.

**Comment by dscho** [comment]
@matteocoder I meant to include you in ⬆️

**Comment by dscho** [comment]
Thank you @ml-evb!

**Comment by dscho** [comment]
Fixed via git-for-windows/msys2-runtime#40

**Comment by albertosantini** [comment]
Thanks @dscho. The snapshot works for me. 
(And I removed the ssh alias, my workaround using the "native" ssh executable).

**Comment by orgads** [comment]
Thank you very much! It works for me too.

**Comment by CrendKing** [comment]
Snapshot works without the `MSYS=enable_pcon` workaround. Looking forward to the release.

**Comment by orgads** [comment]
@dscho I'm using the snpahost version. I ran `git clean -dfx` on a Node.JS project, which has huge `node_modules` directories, that produces a lot of input in high rate, and the terminal froze for very long in the middle of a line, and continued only after the operation completed.

It was more or less:
```
Removing common/schema/
Removing common/setup-ext.js
Removing common/setup-ext.js.map
Removing common/types/bot-activity.js
Removing common/ty
```
\<freeze for about 90s>
```
pes/bot-activity.js.map (continued on the same line)
Removing common/types/common.types.alarm.js
Removing common/types/common.types.alarm.js.map
Removing common/types/common.types.cdr.js
...
```
No output was lost.

Is this expected?

**Comment by dscho** [comment]
> Is this expected?

No. Could you report this to the Cygwin project? It is likely an issue in the Cygwin runtime (as opposed to MSYS2's or Git for Windows' modifications).

**Comment by orgads** [comment]
After I [reported to their mailing list](https://sourceware.org/pipermail/cygwin/2022-February/250852.html) and then I discovered that it doesn't reproduce with Cygwin git, only with Git for Windows.

I'll report it as a new issue.

**Comment by orgads** [comment]
I discovered another issue with the new release. Typed characters are mis-ordered when typing during execution of a foreground job. [Reported to Cygwin](https://sourceware.org/pipermail/cygwin/2022-February/250957.html).

**Comment by dscho** [comment]
> I discovered another issue with the new release. Typed characters are mis-ordered when typing during execution of a foreground job. [Reported to Cygwin](https://sourceware.org/pipermail/cygwin/2022-February/250957.html).

Could you backport https://cygwin.com/pipermail/cygwin-patches/2022q1/011791.html and open a PR at https://github.com/git-for-windows/msys2-runtime/? This will build the MSYS2 runtime as a runtime artifact and you can test whether it fixes the problem you reported. If it does, we can easily merge and build a new `msys2-runtime` package.

**Comment by orgads** [comment]
There was another fix after that. I asked when the next release is planned, but they still haven't replied.

I'll backport both fixes.

**Comment by orgads** [comment]
https://github.com/git-for-windows/msys2-runtime/pull/42

Can you change your base repository to https://github.com/msys2/msys2-runtime? I had to create a new fork for opening the PR.

**Comment by dscho** [comment]
> Can you change your base repository to https://github.com/msys2/msys2-runtime?

Unfortunately, I can't.

**Comment by gdh1995** [comment]
Still meet this with v2.35.1.windows.2. V2.34.0 works.

**Comment by macdjord** [comment]
FYI: The workaround of adding `enable_pcon` to `$MSYS` fixes the use of arrow keys in the pager for me - but it breaks the output of various Git commands, causing them to be printed with weird, deep indentation:
```text
$ git fetch -p
From gitlab.com:redacted/redacted/redacted
                                                 - [deleted]         (none)     -> origin/REDCT-1908-Fix_blank_labels
                                                                                                                     remote: Enumerating objects: 1, done.
                                                                                                                                                          remote: Counting objects: remote: Counting objects: 100% (1/1), done.
                                           remote: Total 1 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (1/1), 300 bytes | 100.00 KiB/s, done.                                         Unpacking objects: 100% (1/1)
                                                                 c091eea..5e54ca1  master       -> origin/master
 + 9083aba...5e54ca1 deployed-DEV -> origin/deployed-DEV  (forced update)
   c091eea..5e54ca1  deployed-QA  -> origin/deployed-QA
```

Version: `git version 2.35.1.windows.2`

**Comment by dscho** [comment]
@macdjord please use the latest [snapshot](https://wingit.blob.core.windows.net/files/index.html) until the next release comes out.

**Comment by jeremyd2019** [comment]
I just updated to 2.35.2.windows.1 and am surprised to still be seeing the arrow keys not working.

**Comment by CrendKing** [comment]
FYI, the latest snapshot (Apr 8) from https://wingit.blob.core.windows.net/files/index.html still works.

**Comment by dscho** [comment]
> I just updated to 2.35.2.windows.1 and am surprised to still be seeing the arrow keys not working in less. I expected to be able to confirm the fix in the 'next release'.

@jeremyd2019 v2.35.2 was a security fix and therefore did not include anything else.

**Comment by rdoering** [comment]
When will it be in the store. (just asking for all unprevileged employees)

**Comment by dscho** [comment]
@rdoering it's not in the store, but if you are asking when will an official Git for Windows be available with this fix: It is now. https://github.com/git-for-windows/git/releases/tag/v2.36.0.windows.1

**Comment by apoetzsch** [comment]
Just upgraded via chocolatey and it fixed the issue. thanks a lot.

**Comment by decatur** [comment]
Migrated to `git version 2.38.1.windows.1` because of a Security Vulnerability.
When ssh-ing, Ctrl-C was then not propagated to the server anymore.
Only workaround for me was `powershell` and `c:/Windows/System32/OpenSSH/ssh.exe`.
So now we have to maintain both a bash  AND a powershell script.
